src/device.rs

```diff
@@ -0,0 +1,96 @@
+// Copyright © 2019 Intel Corporation. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 AND BSD-3-Clause
+
+//! Handles routing to devices in an address space.
+use std::sync::{Arc, Mutex};
+use vm_memory::{GuestAddress, GuestUsize};
+
+/// Trait for devices with basic functions.
+#[allow(unused_variables)]
+pub trait Device: Send {
+    /// Get the device name.
+    fn name(&self) -> String;
+    /// Read from the guest physical address `addr` to `data`.
+    fn read(&mut self, addr: GuestAddress, data: &mut [u8], io_type: IoType);
+    /// Write `data` to the guest physical address `addr`.
+    fn write(&mut self, addr: GuestAddress, data: &[u8], io_type: IoType);
+    /// Set the allocated resource to device.
+    ///
+    /// This will be called by DeviceManager::register_device() to set
+    /// the allocated resource from the vm_allocator back to device.
+    fn set_resources(&mut self, res: &[IoResource], irq: Option<IrqResource>);
```

Comment:
I think this is really to initialize the device instance,  so how about renaming this to something like initialize()/setup() and returning a Result?

Reply:
And feels it would be easier to pass res and irq by value instead of by reference.

Reply:
> I think this is really to initialize the device instance, so how about renaming this to something like initialize()/setup() and returning a Result?

It's more of a notification back to the device than an initialization. The resources have been allocated by the system allocator and it's telling the device about it. It's up to the device to decide what to do with that information.

Reply:
I'm thinking about how to support device hot plug/unplug.
On boot:
1) create a device instance.
2) register the device instance to the device manager.
3) allocate resources for all registered devices.
4) start all registered devices

On hot-add:
1) create device instance
2) register device instance to the device manager
2.1) allocate resources for the device
2.2) start the device

On hot-removal:
1) look up the device
2) stop the device
3) free resources of the device
4) remove the device from the device manager
5) destroy the device

Seems there  are still some missing interfaces to support above flow. 
So I suggested to set_resources() as start(), and add missing stop().

Reply:
> And feels it would be easier to pass res and irq by value instead of by reference.

Could I ask why pass res as Vec<> instead of &[] would be easier/simple?  BTW, this function is called inside register_device(..., resources: &mut Vec<IoResource>)